code/messaging/src/phone/java/com/adobe/marketing/mobile/MessageDelegate.java

```diff
@@ -0,0 +1,240 @@
+        return true;
+    }
+
+    /**
+     * Invoked when a {@link AEPMessage} is attempting to load a URL.
+     *
+     * @param fullscreenMessage the {@link FullscreenMessage} instance
+     * @param urlString         {@link String} containing the URL being loaded by the {@code AEPMessage}
+     * @return true if the SDK wants to handle the URL
+     */
+    @Override
+    public boolean overrideUrlLoad(final FullscreenMessage fullscreenMessage, final String urlString) {
+        Log.trace(LOG_TAG, "%s - Fullscreen overrideUrlLoad callback received with url (%s)", SELF_TAG, urlString);
+
+        if (StringUtils.isNullOrEmpty(urlString)) {
+            Log.debug(LOG_TAG, "%s - Cannot process provided URL string, it is null or empty.", SELF_TAG);
+            return true;
+        }
+
+        URI uri;
+
+        // we need to url encode any javascript if present in the url
+        String localUrlString = urlString;
+        final String[] tokens = urlString.split(AMPERSAND);
+        if (tokens[tokens.length - 1].contains(EXPECTED_JAVASCRIPT_PARAM)) {
+            try {
+                // encode the content after "js="
+                final String urlEncodedJavascript = URLEncoder.encode(tokens[tokens.length - 1].substring(3), StandardCharsets.UTF_8.toString());
+                localUrlString = tokens[0] + AMPERSAND + EXPECTED_JAVASCRIPT_PARAM + urlEncodedJavascript;
+                // the UrlEncoder replaces spaces with "+". we need to manually encode "+" to "%20"".
+                localUrlString = localUrlString.replace("+", "%20");
+            } catch (UnsupportedEncodingException unsupportedEncodingException) {
+                Log.debug(LOG_TAG, "%s - Invalid encoding type (%s), javascript will be ignored.", SELF_TAG, StandardCharsets.UTF_8);
+            }
+        }
+
+        try {
+            uri = new URI(localUrlString);
+        } catch (URISyntaxException ex) {
+            Log.debug(LOG_TAG, "%s - Invalid message URI found (%s).", SELF_TAG, urlString);
+            return true;
+        }
+
+        // check adbinapp scheme
+        final String messageScheme = uri.getScheme();
+
+        if (!messageScheme.equals(MessagingConstants.MESSAGING_SCHEME.ADOBE_INAPP)) {
+            Log.debug(LOG_TAG, "%s - Invalid message scheme found in URI. (%s)", SELF_TAG, urlString);
+            return false;
+        }
+
+        // Populate message data
+        final String query = uri.getQuery();
+        final Map<String, String> messageData = UrlUtilities.extractQueryParameters(query);
+
+        final MessageSettings aepMessageSettings = ((AEPMessage) fullscreenMessage).getSettings();
+        final Message message = (Message) aepMessageSettings.getParent();
+
+        if (messageData != null && !messageData.isEmpty()) {
+            // handle optional tracking
+            final String interaction = messageData.get(MessagingConstants.MESSAGING_SCHEME.INTERACTION);
```

Comment:
Since this function is quite big, I'd suggest to extract few functions from it. For ex: A separate function to handle tracking, a separate function to handle deep link etc.